packages/monitoring-system/data-processor/src/data-processors/cloud-logs-data-processor.ts

```diff
@@ -48,7 +88,341 @@ export class CloudLogsProcessor extends DataProcessor {
     this.listenLimit = options.listenLimit;
   }
 
+  /**
+   * Start the collection and processing task
+   *
+   * @returns a promise that settles once all received messages
+   * have been processed (successfully or unsuccessfully). The promise
+   * will be fulfilled even if some messages are unsuccessfully processed.
+   * The promise will only reject for runtime errors such as loss of
+   * communication to PubSub.
+   */
   public async collectAndProcess(): Promise<void> {
-    throw new Error('Method not implemented.');
+    this.subscription.on('message', this.processMessage.bind(this));
+    this.logger.debug('Now listening for PubSub messages');
+
+    return new Promise((resolve, reject) => {
+      const stopListening = () => {
+        this.subscription.removeAllListeners();
+        this.logger.debug(`Stopped listening after ${this.listenLimit}s`);
+
+        Promise.all(this.tasksInProgress)
```

Comment:
Does this check every message in the list or does it short-circuit?

Reply:
It short-circuits (i.e. `rejects` with the first rejected promise in the array) but recall that we only `reject` promises for runtime errors (eg. Pub/Sub connectivity issues or Firestore issues) and not for issues related to individual messages (eg. a malformed log entry).

So it makes sense to short-circuit on the first rejection because for runtime errors it's highly probable that the other promises down the line will face the same issue and just throw errors.